### ISSUE TITLE
Regression in memory allocs when forming NTuple 

### ISSUE THREAD

**Issue by jw3126** [JuliaLang/julia#41512]
The following code allocates on master and 1.7beta2, but does not allocate on 1.6:
```julia
Tuple3(itr) = NTuple{3, Float64}(itr)::NTuple{3,Float64}

function doit(itr)
    for x in itr
        Tuple3(x)
    end
end

itr = [randn(3) for _ in 1:10000]
doit(itr)
@time doit(itr)
```
```
master:  0.021418 seconds (120.00 k allocations: 2.899 MiB)
julia 1.6:  0.000032 seconds
```
This lead to regressions in two packages for me: 
https://github.com/jw3126/VoxelRayTracers.jl/blob/master/src/core.jl#L71
https://github.com/jw3126/LinearInterpolations.jl/blob/master/src/LinearInterpolations.jl#L407
https://github.com/jw3126/LinearInterpolations.jl/blob/master/test/test_LinearInterpolations.jl#L259

**Comment by simeonschaub** [comment]
Fixed in #41074, but looks like we forgot to mark that for backporting.

**Comment by jw3126** [comment]
@simeonschaub I could reproduce with today's master. Note that type inference works, but still it allocates for me. Could you run my example just to double check?

**Comment by jw3126** [comment]
@simeonschaub this is how it looks on today's master for me:
```julia
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.8.0-DEV.155 (2021-07-07)
 _/ |\__'_|_|_|\__'_|  |  Commit 97f817a379* (0 days old master)
|__/                   |

julia> Tuple3(itr) = NTuple{3, Float64}(itr)::NTuple{3,Float64}
Tuple3 (generic function with 1 method)

julia> function doit(itr);
           for x in itr
               Tuple3(x)
           end
       end
doit (generic function with 1 method)

julia> itr = [randn(3) for _ in 1:10000];doit(itr);

julia> @time doit(itr)
  0.023238 seconds (120.00 k allocations: 2.899 MiB)
```

**Comment by simeonschaub** [comment]
Ah sorry, I misunderstood. Would be good to fix this.

**Comment by JeffBezanson** [comment]
Base._totuple was inlined in 1.6, but isn't anymore. The cost is fine, and the method is even marked inline, so I guess it might be due to recursion detection?

**Comment by jw3126** [comment]
Seems to be fixed for me on 1.8.5.
